### ISSUE TITLE
tune: support for conditional / nested, hierarchical) parameter search spaces

### ISSUE THREAD

**Issue by bbudescu** [microsoft/FLAML#816]
This is not documented, so I assume that, most likely, it's not directly supported. However, I see that there is a discussion about it in #

**Comment by bbudescu** [comment]
This is not documented, so I assume that, most likely, it's not directly supported. However, I see that there is a discussion about it in #717.

Also, I think the constraints [api](https://microsoft.github.io/FLAML/docs/Use-Cases/Tune-User-Defined-Function#more-constraints-on-the-tuning) can be used to mark incompatible configurations as invalid. I'm not sure, though, how that affects tuning performance. I haven't taken the time to properly understand the [BlendSearch](https://www.microsoft.com/en-us/research/publication/economical-hyperparameter-optimization-with-blended-search-strategy/) algorithm and how it trades off local with global search, but I imagine that the bayesian model underlying at least the global search (that must be a Gaussian Process, or something, right?), that serves as a predictor for the performance which would be attained by a certain config before running it, would have some sort of smoothness prior. Now, if using constraints assigns very low scores to invalid configurations, due to this smoothness prior, wouldn't it also assign low scores probs to the configs at the edge of feasibility? I imagine that there might be very good combinations right at the edge of the feasibility region, but they will be explored later rather than sooner, and, although kinks can be modeled, they are still considered less likely than smooth regions.

So, in short, is it safe to use the relational constraints to implement conditional search spaces? Is there another, better way to let the optimizer know that a particular combo is not feasible?

Also, I know that some other black box optimization packages, e.g., [HyperMapper](https://github.com/luinardi/hypermapper/wiki/Chakong%E2%80%90Haimes#the-json-configuration-file) train a separate classifier model for guessing whether a certain config is feasible or not. Do you guys use such a model to drive sampling, too?

I mean, it has the advantage that you don't have to specify the invalid space up front, which is nice if you don't know the invalid combinations _a priori_, and you want the optimizer to learn this by itself, but this is not very helpful if you do have the invalid combos.

**Comment by bbudescu** [comment]
Is there perhaps an example of how to use hierarchical / nested search spaces that I can use until you guys are done with the docs?

**Comment by sonichi** [comment]
> Is there perhaps an example of how to use hierarchical / nested search spaces that I can use until you guys are done with the docs?

hierarchical: https://github.com/microsoft/FLAML/blob/5eb99276423f8ac88b7fe38c37d9b7f37fafe4fb/test/tune/test_space.py#L47
nested:
https://github.com/microsoft/FLAML/blob/5eb99276423f8ac88b7fe38c37d9b7f37fafe4fb/test/tune/test_tune.py#L221

**Comment by bbudescu** [comment]
how about more advanced conditional search spaces than a plain choice (e.g., similar to #988)?
